jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/ExtractTarStep.java

```diff
@@ -0,0 +1,147 @@
+      this.baseImage = baseImage;
+      this.layers = layers;
+    }
+  }
+
+  /**
+   * Checks the first two bytes of a file to see if it has been gzipped.
+   *
+   * @param path the file to check
+   * @return {@code true} if the file is gzipped, {@code false} if not
+   * @throws IOException if reading the file fails
+   * @see <a href="http://www.zlib.org/rfc-gzip.html#file-format">GZIP file format</a>
+   */
+  @VisibleForTesting
+  static boolean isGzipped(Path path) throws IOException {
+    try (InputStream inputStream = Files.newInputStream(path)) {
+      inputStream.mark(2);
+      int magic = (inputStream.read() & 0xff) | ((inputStream.read() << 8) & 0xff00);
+      return magic == GZIPInputStream.GZIP_MAGIC;
+    }
+  }
+
+  private final Path tarPath;
+  private final Path destination;
+
+  ExtractTarStep(Path tarPath, Path destination) {
+    this.tarPath = tarPath;
+    this.destination = destination;
+  }
+
+  @Override
+  public LocalImage call()
+      throws IOException, LayerCountMismatchException, BadContainerConfigurationFormatException {
+    TarExtractor.extract(tarPath, destination);
+
+    InputStream manifestStream = Files.newInputStream(destination.resolve("manifest.json"));
+    DockerManifestEntryTemplate loadManifest =
+        new ObjectMapper()
+            .configure(MapperFeature.ACCEPT_CASE_INSENSITIVE_PROPERTIES, true)
+            .readValue(manifestStream, DockerManifestEntryTemplate[].class)[0];
+    manifestStream.close();
+    ContainerConfigurationTemplate configuration =
+        JsonTemplateMapper.readJsonFromFile(
+            destination.resolve(loadManifest.getConfig()), ContainerConfigurationTemplate.class);
+
+    List<String> layerFiles = loadManifest.getLayerFiles();
+    if (configuration.getLayerCount() != layerFiles.size()) {
+      throw new LayerCountMismatchException(
+          "Invalid base image format: manifest contains "
+              + layerFiles.size()
+              + " layers, but container configuration contains "
+              + configuration.getLayerCount()
+              + " layers");
+    }
+
+    // Check the first layer to see if the layers are compressed already. 'docker save' output is
+    // uncompressed, but a jib-built tar has compressed layers.
+    boolean layersAreCompressed =
+        layerFiles.size() > 0 && isGzipped(destination.resolve(layerFiles.get(0)));
+
+    // Process layer blobs
+    // TODO: Optimize; compressing/calculating layer digests is slow
+    List<PreparedLayer> layers = new ArrayList<>();
+    V22ManifestTemplate v22Manifest = new V22ManifestTemplate();
+    for (int index = 0; index < layerFiles.size(); index++) {
+      Path file = destination.resolve(layerFiles.get(index));
+
+      // Compress layers if necessary and calculate the digest/size
+      Blob blob = layersAreCompressed ? Blobs.from(file) : Blobs.compress(Blobs.from(file));
+      BlobDescriptor blobDescriptor = blob.writeTo(ByteStreams.nullOutputStream());
```

Comment:
Potential optimization idea:

So, we end up writing each (compressed) blob content (layer) two times: first to compute the layer digest (which is required to assemble the manifest) and secondly to actually push to the registry. (I may be wrong about this though.) We can push (we currently do) the manifest as the last step, so I think theoretically it's possible that you start pushing base image layers early while compressing on-the-fly and this will get us their digests. This way, we may be able to write compressed contents once.

But I may be missing something in this process. Can't remember all the details.

Reply:
This sounds like it may be possible, but would probably require more build step refactoring, so I'll hold off on it for now.

Reply:
Yeah, let's just keep this in mind. This, and also the idea to untar the output of `docker save` on-the-fly and skip writing the tar onto the disk. Oh, and the possibility to reuse the tar (untarred or not), if we can correctly notice changes. Maybe file an issue to track regardless of how difficult they are?

Reply:
I'll open one for the caching ideas and another for the other optimizations (on-the-fly untar, compression while pushing) after this gets merged.